src/stepFunctions/commands/visualizeStateMachine/aslVisualization.ts

```diff
@@ -153,16 +153,28 @@ export class AslVisualization {
         )
 
         // When the panel is closed, dispose of any disposables/remove subscriptions
-        panel.onDidDispose(() => {
+        const disposePanel = () => {
             debouncedUpdate.cancel()
             this.onVisualizationDisposeEmitter.fire()
             this.disposables.forEach(disposable => {
                 disposable.dispose()
             })
             this.onVisualizationDisposeEmitter.dispose()
             this.isPanelDisposed = true
+        }
+
+        panel.onDidDispose(() => {
+            if (!this.isPanelDisposed) {
+                disposePanel()
+            }
         })
 
+        panel.dispose = () => {
```

Comment:
Overwriting the `WebviewPanel` dispose method seems risky. Is `onDidDispose` sufficient?

Reply:
The original dispose() could at least be preserved like:

```
const parentDispose = panel.dispose
panel.dispose = () => {
  ...
  parentDispose()
}
```

Reply:
@awschristou the original problem was caused by a time delay between `dispose` and `onDidDipose` (async). Overwriting `dispose` and then calling the original impl ensures there is no delay.

Reply:
Learning is that you can't start a background task and rely on `onDidDispose` to dispose it because there's no facility to tell in the background test whether you've already been disposed.

This could be addressed by VSCode exposing [the `isDisposed` flag on the webview](https://github.com/microsoft/vscode/blob/4862602c4c3a14de76e24898c4d94a79210f7356/src/vs/workbench/api/common/extHostWebview.ts#L43). The workaround here is to recreate such a flag and make sure it reflects the status of `dispose`.

Reply:
Would https://github.com/aws/aws-toolkit-vscode/pull/1098#discussion_r427464003 help with that? Messing with a third party type, specifically the VS Code API opens the door to incompatibility problems.

Reply:
No since the async `debouncedUpdate` can still be executed before the async `onDidDipose`.

The bad ordering is `dispose()` => `debouncedUpdate()` => `onDidDispose()`

Reply:
What is the core problem? Is it that sendUpdateMessage (debounced) is running when the view is disposed? Sounds like handling may be needed in other places to address the cause instead of the symptoms.

Reply:
Fixed the first test failure. The second and third test failures are caused by the call to parentDispose(). If I comment out the call to parentDispose() the tests all pass without error and the extension works as intended. Any idea on why it produces "TypeError: attempted to get private field on non-instance"?

Reply:
Is there a race condition going on with respect to when other events are running and when the object is disposed? It should be possible to determine where this is coming up and assess if enough checks are in place. One idea: emulating an onWillDispose (prior to allowing the actual dispose call to proceed) would allow you to set a flag for the rest of the code in the class to check, then onDidDispose can shut things down.